`Communication`: Allow users to view and download sent attachments

julian-wls opened the pull request:
<!-- Feel free to leave out sections that are not appropriate for your PR.  -->

### Problem Description

As described in #117 all non-image attachments that are sent in the chat are rendered as links and cannot be opened currently due to an authorization error, since there is no authToken attached to the request.

### Changes
<!-- Descripe your changes on a high level. If you feel like technical details would be helpful for the reviewer, please add them here. --> 

This PR introduces a Link resolver, that is capable of detecting the target of the link and attaches an authToken if needed. If the link targets a pdf file, the resolver passes the request to a new Pdf View, which is wrapped in a modal bottom sheet. Requests that target other Artemis pages should also be opened in a modal bottom sheet in a web view, which allows attaching cookies, however there were issues (see #245). All other links are now opened as new IntentTab.
The new Pdf view was inspired by [this repository](https://github.com/GRizzi91/bouquet). It allows users to share and download the pdf. Pdf can also be shown vertically and horizontally.

Note: The security checks have failed because of [the repository](https://github.com/GRizzi91/bouquet) being temporarily added to the project. However, it has been removed and the pdf viewer was added manually including various modifications (see *pdf* package in `core:ui` module)

### Steps for testing

1. Go into a chat that contains a resource pdf.
2. Click on it and verify, that the pdf view opens the pdf.
3. Try downloading, sharing and rotating.
4. Turn of your internet connection and try loading the pdf again, verify that an appropriate error message is shown.

### Screenshots
![PdfViewer](https://github.com/user-attachments/assets/7737c45c-11da-45ca-9b37-ce815d367f4e)

FelberMartin reviewed (changes requested):
Looking very good overall, I think this is really a major improvement for the app. Good job! 😊 

Besides the code comments, I have a few general remarks:
1. When opening the pdf, it shows a page count of "Page 3/18", even though only the first 1,5 pages are visible. But I think you mentioned something about the page indicator behaving weirdly, so if there is no easy fix, its also fine if we leave it as is :)
2. The name of the pdf is currently just the timestamp (if think), could we change this to the name displayed in the markdown link? When I eg download the slides I cant really judge what the content of the file is by its name.
3. The action labelled with "Rotate PDF" felt quite counter-intuitive for me, from this name I would have expected the pdfs to rotate 90°. But I also can't find a short, better fitting alternative. Its a little bit longer than the other labels, but maybe "Toggle View Mode" would be a better fit, what do you think @julian-wls ?
4. Somewhere in the code zoom is mentioned, I tried to zoom the pdfs by pinching with two fingers in the bottomSheet, but nothing happened. Is this intended?

![image](https://github.com/user-attachments/assets/98734a0b-0761-448e-afe6-57adbdbf52b5)

julian-wls commented on core/ui/src/main/kotlin/de/tum/informatics/www1/artemis/native_app/core/ui/markdown/link_resolving/MarkdownLinkResolverImpl.kt:
```diff
@@ -0,0 +1,80 @@
+//            }
+            else -> {
+                val customTabsIntent = CustomTabsIntent.Builder().build()
+                customTabsIntent.launchUrl(context, Uri.parse(link))
```
Oh true, I totally forgot that.

julian-wls commented on core/ui/src/main/kotlin/de/tum/informatics/www1/artemis/native_app/core/ui/markdown/link_resolving/MarkdownLinkResolverImpl.kt:
```diff
@@ -0,0 +1,80 @@
+        val context = LocalContext.current
+
+        val (bottomSheetLink, setLinkToShow) = remember { mutableStateOf<String?>(null) }
+        val (bottomSheetState, setBottomSheetState) = remember { mutableStateOf(LinkBottomSheetState.WEBVIEWSTATE) }
```
The main reason was, that I had a hard time debugging it at some point, cause the code was not quite readable before. That's the reason I changed it to this syntax.

julian-wls commented:
> * When opening the pdf, it shows a page count of "Page 3/18", even though only the first 1,5 pages are visible. But I think you mentioned something about the page indicator behaving weirdly, so if there is no easy fix, its also fine if we leave it as is :)
> * The name of the pdf is currently just the timestamp (if think), could we change this to the name displayed in the markdown link? When I eg download the slides I cant really judge what the content of the file is by its name.
> * The action labelled with "Rotate PDF" felt quite counter-intuitive for me, from this name I would have expected the pdfs to rotate 90°. But I also can't find a short, better fitting alternative. Its a little bit longer than the other labels, but maybe "Toggle View Mode" would be a better fit, what do you think @julian-wls ?
> * Somewhere in the code zoom is mentioned, I tried to zoom the pdfs by pinching with two fingers in the bottomSheet, but nothing happened. Is this intended?

Thanks for reviewing. I really appreciate the feedback :)

Here are the answers:
1. It actually shows three pages when you open the entire modal bottom sheet. There is a function in VerticalPdfReaderState, which determines how many pages are shown and sets the current page to the last page that is currently open. If we change this the last page of the pdf would never be reached.
2. I thought about this too, but I didn't find any way to access the string in the markdown text (I think that would be the only way to get the name). The Link resolver only checks when a link has been clicked and only returns the link, not the title.
3. I think you're right. Toggle view mode sounds good to me. Would you still keep this icon, though? @FelberMartin 
4. For zooming you have to double tap the page, first.

eylulnc commented:
Thanks for the great work! This is indeed a huge improvement 😊

I checked it out based on our previous discussions and have a few observations:
**Page Indicator:** Regarding the ‘Page 3/18’ indicator, I didn’t encounter this issue during my testing—it seems to be working fine for me.
**Bottom Sheet Behavior:** I really like the toast—it’s a nice touch! However, I’d prefer not to see an empty bottom sheet when there’s an error (e.g., no internet connection), as it feels a bit disconnected.

<img width="200" alt="Screenshot 2024-12-22 at 15 58 38" src="https://github.com/user-attachments/assets/7622dc02-49ff-4477-b49c-9bf12d591379" />

Additionally, I noticed the bottom sheet expands into the unsafe zone—perhaps adding some padding at the top could make it look better. What do you think?

**Naming:**  I agree it would be nice to have more meaningful names for downloaded files. However, if this isn’t feasible, it’s fine as is.

<img width="200" alt="Screenshot 2024-12-22 at 16 00 05" src="https://github.com/user-attachments/assets/161b0408-bbcf-458e-b3b0-13a03d969c28" />

**Rotation Action:** Initially, I found the label “Rotate PDF” a bit confusing because I expected it to change the PDF’s orientation. Since it toggles between scroll and swipe modes, maybe a label like “Read Mode” could help clarify this. Keeping the current icon is fine, but if you feel the label doesn’t cause confusion, I think it’s okay to keep it as is.

FelberMartin reviewed (approved):
Lgtm now! 

Regarding the naming of the "Rotate" item, I thinks its worth to take the bigger dialog to avoid the confusing naming of "Rotate pdf". Keeping the icon seems fine to me

FelberMartin commented:
> Bottom Sheet Behavior: I really like the toast—it’s a nice touch! However, I’d prefer not to see an empty bottom sheet when there’s an error (e.g., no internet connection), as it feels a bit disconnected.

@julian-wls did you consider this comment by Eylül?

julian-wls commented:
> > Bottom Sheet Behavior: I really like the toast—it’s a nice touch! However, I’d prefer not to see an empty bottom sheet when there’s an error (e.g., no internet connection), as it feels a bit disconnected.
> 
> @julian-wls did you consider this comment by Eylül?

Sorry I overlooked it, thanks for the hint. I added a change that allows closing the bottom sheet in case an error occurred and added padding to the top.
